### PR TITLE
Synchronise multiple signal

### PR DESCRIPTION
**Pull request by Edouard2laire**
Hello, 

This process aims at aligning multiple signals acquired with different devices but sharing synchronizing events (typically sent to each recording device at the time of the acquisition).  This situation can arise easily when recording multimodal data such as EEG - iEEG,  EEG - fMRI, fNIRS - fMRI, or EEG - fMRI. 

Having the signal synchronised makes making easier to perform multimodal analysis, notably allowing the simultaneous exploration of the different signals. For example, here is a figure I recently presented at a conference, showing how we can understand better what is happening during a seizure by looking at the EEG, fNIRS, and pulse oximeter 

image

What's left to do: 

- [x]  Fix the code to save continuous files as output 
- [x]  Improve file history

Future work 

Here is a list of future things we might consider implementing but that is not required for this PR

- [x]  Allowing the simultaneous visualization of the different signals from Brainstorm (#656)
- [x]  Explicitly linking the different synchronized files. This could notably be used so that if an event is added to a file, then the same event would be added to all the events the file is synchronized with. 
- [ ]  Add the option to either crop the longer signal  keeping only the common signal (current implementation) or pad the shorter signal with either 0 or NaN

Let me know if you want me to send you an example dataset. I can generate a minimum brainstorm database with 2 files to synchronize. 

Regards, 
Edouard

### CODE REVIEW

File: toolbox/process/functions/process_synchronise_multiples_signals.m [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    
+    % === Sync event management === %
+    SyncEventName  = sProcess.options.src.Value;
+    
+    sData           = cell(1,length(sInputs)); 
+    sDataRaw        = cell(1,length(sInputs)); 
+    sSync           = cell(1,length(sInputs)); 
+    fs              = zeros(1, length(sInputs)); 
+    
+    is_raw = strcmp(sInputs(1).FileType, 'raw');
```
Comment (Edouard2laire) [inline toolbox/process/functions/process_synchronise_multiples_signals.m:69]: Maybe we would like to ensure that all signal are either raw or imported data

File: toolbox/process/functions/process_synchronise_multiples_signals.m [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            sFileIn.header.nsamples = length( new_data{iFile}.Time );
+            sFileIn.prop.times      = [ new_data{iFile}.Time(1),  new_data{iFile}.Time(end)];
+            [sFileOut, errMsg] = out_fopen(RawFileOut, 'BST-BIN', sFileIn, ChannelMat);
+            
+            % Set Output sFile structure
+            sOutMat.format = 'BST-BIN';
+            sOutMat.F = sFileOut;
+            sOutMat.DataType        = 'raw'; 
+            sOutMat.History         = new_data{iFile}.History;
+            sOutMat                 = bst_history('add', sOutMat, 'process', 'Synchronisation');
```
Comment (Edouard2laire) [inline toolbox/process/functions/process_synchronise_multiples_signals.m:223]: Might be improved listing the synchronized files and event names used for synchronization

File: toolbox/process/functions/process_synchronise_multiples_signals.m [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        new_data{iFile}.F       = sDataTmp.F(:,index);
+        
+        if ~is_raw
+            tmp_event = new_data{iFile}.Events;
+        else
+            tmp_event = sDataRaw{iFile}.F.events;
+        end
+
+        for i_event = 1:length(tmp_event)
+            tmp_event(i_event).times = tmp_event(i_event).times - mean_shifting(iFile) - new_times{iFile}(index(1)) ;
```
Comment (Edouard2laire) [inline toolbox/process/functions/process_synchronise_multiples_signals.m:159]: we might want to check that the new event is not outside the time definition of the file

File: toolbox/process/functions/process_synchronise_multiples_signals.m (hunk 1 above)
Comment (Edouard2laire) [reply toolbox/process/functions/process_synchronise_multiples_signals.m]: done :)

**Comment by rcassani** [comment]
> Let me know if you want me to send you an example dataset. I can generate a minimum brainstorm database with 2 files to synchronize.

@Edouard2laire, yes, please send me a minimal example dataset

**Comment by Edouard2laire** [comment]
Hi. 
DId you had time to review this PR and [the](https://github.com/brainstorm-tools/brainstorm3/issues/656)  @rcassani  ? 

Thanks a lot, 
Edouard

**Comment by rcassani** [comment]
@Edouard2laire, apologies for the delay. there was a misunderstanding, I thought the PR still was on development. I will review it ASAP

File: toolbox/process/functions/process_synchronise_multiples_signals.m (hunk 3 above)
Comment (rcassani) [reply toolbox/process/functions/process_synchronise_multiples_signals.m]: Handled at [de3c5bd](https://github.com/brainstorm-tools/brainstorm3/pull/660/commits/de3c5bd2e5ee74f396aa31f777692c5d1d3e5b7c)

File: toolbox/process/functions/process_synchronise_multiples_signals.m (hunk 2 above)
Comment (rcassani) [reply toolbox/process/functions/process_synchronise_multiples_signals.m]: Added at [b10443c](https://github.com/brainstorm-tools/brainstorm3/pull/660/commits/b10443cc78b48f7cf3afc91bb9f99c5882852c34)

**Comment by rcassani** [comment]
@Edouard2laire, the main change in the process is that files are loaded one by one in memory.

Minor changes: 
* It now removes the completely out-of-range events, and fixes the partially out extended events.
* List of synch files in History

File: toolbox/process/functions/process_synchronise_multiples_signals.m [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    if any(~(strcmp({sEvtSync.label}, syncEventName)))
+        bst_error(['Sync event ("' syncEventName '") must be present in all files'], 'Synchronize signal', 0);
+        return;
+    end
+
+    % Check: Sync event must be present in all files
+    if any(cellfun(@(x) size(x,1), {sEvtSync.times}) ~= 1)
+        bst_error(['Sync event ("' syncEventName '") must be simple event in all the files'], 'Synchronize signal', 0);
+        return;
+    end
```
Comment (rcassani) [inline toolbox/process/functions/process_synchronise_multiples_signals.m:109]: For now it limits the synch event to be simple (not extended). 
Do you see any value in adding support for extended processes?

File: toolbox/process/functions/process_synchronise_multiples_signals.m
[the remark this answers is not in the document]
Comment (Edouard2laire) [reply toolbox/process/functions/process_synchronise_multiples_signals.m:1]: sounds good to me.

File: toolbox/process/functions/process_synchronise_multiples_signals.m (hunk 4 above)
Comment (Edouard2laire) [reply toolbox/process/functions/process_synchronise_multiples_signals.m:109]: i dont think that could be useful also might add more complication than necessary.

File: toolbox/process/functions/process_synchronise_multiples_signals.m [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+    bst_progress('inc', nInputs);
+    bst_progress('text', 'Synchronizing...');
+
```
[the remark this answers is not in the document]
Comment (Edouard2laire) [reply toolbox/process/functions/process_synchronise_multiples_signals.m:113]: indeed, i think having the first file with the highest sampling freq might be better in this case. so far I was always putting the EEG first when synchronizing. 

Should we do the sorting in the process ? it's adding a bit of code to get the sampling rate of each file and then sort them.

**Comment by Edouard2laire** [comment]
Thanks a lot for all the changes! 
I'll test it this weekend and let you know if everything works. 

Edouard

File: toolbox/process/functions/process_synchronise_multiples_signals.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/process/functions/process_synchronise_multiples_signals.m]: Added in [5e1dd72](https://github.com/brainstorm-tools/brainstorm3/pull/660/commits/5e1dd72883ab92333b3d0337180c3a037ec7aa29)

File: toolbox/process/functions/process_synchronise_multiples_signals.m
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/process/functions/process_synchronise_multiples_signals.m:1]: Done in [cd7e8b5](https://github.com/brainstorm-tools/brainstorm3/pull/660/commits/cd7e8b5106b6683a5df2bec37d61c74c897ee618)

**Comment by Edouard2laire** [comment]
Hello. 

I did few test and it seems to be working. Thanks a lot, 

Edouard

**Comment by rcassani** [comment]
@Edouard2laire, great!

**Comment by Edouard2laire** [comment]
(Clicked wrong button — you can ignore that)

**Comment by rcassani** [comment]
For the open tasks:
- The feature of multiple raw time-series plots can be handled in another PR.
- Signals are now chopped to be the same length, not sure if the alternative of zero-padding would be useful
- Synching other events (after synching signals) for the moment can be done with [Transfer events](https://neuroimage.usc.edu/brainstorm/Tutorials/EyetrackSynchro#Event_transfer)

**Comment by Edouard2laire** [comment]
> For the open tasks:
> 
> * The feature of multiple raw time-series plots can be handled in another PR.

Do you want me to create the PR ? I don't really have any idea on where to start for that implementation. 


> * Signals are now chopped to be the same length, not sure if the alternative of zero-padding would be useful

yes, I agree that it's not really useful.  Could be implemented if someone asks. This can be the case if you still want to keep all the signals and don't have the constraint of having all the recordings all the time. 


> * Synching other events (after synching signals) for the moment can be done with [Transfer events](https://neuroimage.usc.edu/brainstorm/Tutorials/EyetrackSynchro#Event_transfer)

yes, that's true

**Comment by rcassani** [comment]
> Do you want me to create the PR ? I don't really have any idea on where to start for that implementation.

I'll start the PR, I need to check some implementation of the raw-file readers

**Pull request merged**